OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,265 @@
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Meebey.SmartIrc4net;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	class IrcLogic
+	{
+		IrcClient client;
+
+		TextFieldWidget inputBox;
+		TextFieldWidget nicknameBox;
+		Widget connectBG;
+		Widget ircContainer;
+
+		ScrollPanelWidget historyPanel;
+		LabelWidget historyTemplate;
+		ScrollPanelWidget nicknamePanel;
+		LabelWidget nicknameTemplate;
+
+		[ObjectCreator.UseCtor]
+		public IrcLogic(Widget widget)
+		{
+			Log.AddChannel("irc", "irc.log");
+
+			historyPanel = widget.Get<ScrollPanelWidget>("HISTORY_PANEL");
+			historyTemplate = widget.Get<LabelWidget>("HISTORY_TEMPLATE");
+			nicknamePanel = widget.Get<ScrollPanelWidget>("NICKNAME_PANEL");
+			nicknameTemplate = widget.Get<LabelWidget>("NICKNAME_TEMPLATE");
+
+			inputBox = widget.Get<TextFieldWidget>("INPUT_BOX");
+			inputBox.OnEnterKey = EnterPressed;
+			inputBox.IsDisabled = () => client == null;
+
+			if (Game.Settings.Irc.Nickname == new IrcSettings().Nickname)
+				Game.Settings.Irc.Nickname += Game.CosmeticRandom.Next(100, 999);
+
+			nicknameBox = widget.Get<TextFieldWidget>("NICKNAME_BOX");
+			nicknameBox.Text = Game.Settings.Irc.Nickname;
+			nicknameBox.OnTextEdited = () => { Game.Settings.Irc.Nickname = nicknameBox.Text; Game.Settings.Save(); };
+
+			connectBG = widget.Get("IRC_CONNECT_BG");
+			ircContainer = widget.Get("IRC_CONTAINER");
+
+			var disconnectButton = widget.Get<ButtonWidget>("DISCONNECT_BUTTON");
+			disconnectButton.IsDisabled = () => client == null;
+			disconnectButton.OnClick = Disconnect;
+
+			MaybeShowConnectPanel();
+		}
+
+		void MaybeShowConnectPanel()
+		{
+			if (client != null && client.IsConnected)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				return;
+			}
+
+			if (Game.Settings.Irc.ConnectAutomatically)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				Connect();
+				return;
+			}
+
+			ircContainer.Visible = false;
+			connectBG.Visible = true;
+
+			var connectAutomaticallyCheckBox = connectBG.Get<CheckboxWidget>("CONNECT_AUTOMATICALLY_CHECKBOX");
+			connectAutomaticallyCheckBox.IsChecked = () => Game.Settings.Irc.ConnectAutomatically;
+			connectAutomaticallyCheckBox.OnClick = () => Game.Settings.Irc.ConnectAutomatically ^= true;
+
+			var connectButton = connectBG.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.OnClick = () =>
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Game.Settings.Irc.ConnectAutomatically = connectAutomaticallyCheckBox.IsChecked();
+				Game.Settings.Save();
+				Connect();
+			};
+		}
+
+		void Connect()
+		{
+			client = new IrcClient();
+			client.Encoding = System.Text.Encoding.UTF8;
+			client.SendDelay = 100;
+			client.ActiveChannelSyncing = true;
+
+			client.OnConnected += new EventHandler(OnConnected);
```

Comment:
This can just be `client.OnConnected += OnConnected;`
Similar for the other lines below